tests/utils/test_instance_grads.py

```diff
@@ -0,0 +1,68 @@
+#%%
+
+from collections import defaultdict
+from typing import Dict, List
+
+import torch as t
+from transformer_lens import HookedTransformer
+
+from auto_circuit.tasks import Task
+from auto_circuit.types import AblationType
+from auto_circuit.utils.ablation_activations import src_ablations
+from auto_circuit.utils.graph_utils import (
+    patch_mode,
+    set_all_masks,
+    set_mask_batch_size,
+    train_mask_mode,
+)
+from auto_circuit.utils.tensor_ops import batch_avg_answer_diff
+
+
+def test_instance_grads(
+    mini_tl_transformer: HookedTransformer, dataset_name: str = "mini_prompts"
+):
+
+    # create task
+    batch_size = 2
+    task = Task(
+        key="test_eap",
+        name="test_eap",
+        batch_size=batch_size,
+        batch_count=1,
+        token_circuit=False,
+        _model_def=mini_tl_transformer,
+        _dataset_name=dataset_name,
+    )
+    model = task.model
+    train_loader = task.train_loader
+
+    # compute patch outs
+    patch_outs = {}
+    for batch in train_loader:
+        patch_outs[batch.key] = src_ablations(
+            model, batch.clean, ablation_type=AblationType.ZERO
+        )
```

Comment:
patch_outs is the same for all batches as this is zero ablation, so this loop is not required